### ISSUE TITLE
Make the RegisterForReflection annotation more flexible for native image

### ISSUE THREAD

**Issue by lbernardomaia** [quarkusio/quarkus#3883]
**Description**
Would be nice have a way to tell to `@RegisterForReflection` annotation what I want instead of using the reflection-config.json file as stated here https://quarkus.io/guides/writing-native-applications-tips.

Currently, If I want only public methods to use reflection, I need to create the file below:

reflection-config.json:
```
[
  {
    "name" : "something.test.resource",
    "allPublicMethods" : true
  }
]
```

Implementation ideas:
`@RegisterForReflection(allPublicMethods=true)`

**Comment by gsmet** [comment]
@lbernardomaia interested in contributing it? It shouldn't be too hard.

Better suggest a design for the annotation before though.

**Comment by lbernardomaia** [comment]
Sure, I had a look and one suggested design would be as below:

RegisterForReflection:
```
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface RegisterForReflection {

    /**
     * If the methods should be registered
     */
    boolean methods() default true;

    /**
     * If the fields should be registered
     */
    boolean fields() default true;

        Attributes[] enableAttributes() default {};

    enum Attributes {
        ALL_DECLARED_CONSTRUCTOR,
        ALL_PUBLIC_CONSTRUCTORS,
        ALL_DECLARED_METHODS,
        ALL_PUBLIC_METHODS,
        ALL_DECLARED_CLASSES,
        ALL_PUBLIC_CLASSES;
    }
}
```

Use:

- @RegisterForReflection(enableAttributes = Attributes.ALL_DECLARED_CLASSES)

- @RegisterForReflection(enableAttributes = {Attributes.ALL_DECLARED_CLASSES, Attributes.ALL_DECLARED_CONSTRUCTOR})

The use of `enableAttributes()` would set the `methods()` and `fields()` to false.

**Comment by lbernardomaia** [comment]
Sure, I had a look and one suggested design would be as below:

RegisterForReflection:
```
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface RegisterForReflection {

    /**
     * If the methods should be registered
     */
    boolean methods() default true;

    /**
     * If the fields should be registered
     */
    boolean fields() default true;

        Attributes[] enabledAttributes() default {};

    enum Attributes {
        ALL_DECLARED_CONSTRUCTOR,
        ALL_PUBLIC_CONSTRUCTORS,
        ALL_DECLARED_METHODS,
        ALL_PUBLIC_METHODS,
        ALL_DECLARED_CLASSES,
        ALL_PUBLIC_CLASSES;
    }
}
```

Use:

- @RegisterForReflection(enabledAttributes = Attributes.ALL_DECLARED_CLASSES)

- @RegisterForReflection(enabledAttributes = {Attributes.ALL_DECLARED_CLASSES, Attributes.ALL_DECLARED_CONSTRUCTOR})

The use of `enabledAttributes()` would set the `methods()` and `fields()` to false.

**Comment by gastaldi** [comment]
@lbernardomaia your proposal seems a bit confusing, shouldn't `methods()` and `fields()` instead of being `boolean` accept a format like in https://github.com/oracle/graal/blob/master/substratevm/REFLECTION.md#manual-configuration or am I missing something?

**Comment by gastaldi** [comment]
Ah that is part of the current annotation, just realized that, sorry. :) 
I think in this case those booleans need to go in favor of your `enabledAttributes` array WDYT?

**Comment by lbernardomaia** [comment]
Agree, but that would be a breaking change for the previous version. Is that a problem?

Removing the those booleans, the class would be:

`
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface RegisterForReflection {

    Attributes[] enableAttributes() default {Attributes.ALL};

    enum Attributes {
        ALL,
        ALL_DECLARED_CONSTRUCTOR,
        ALL_PUBLIC_CONSTRUCTORS,
        ALL_DECLARED_METHODS,
        ALL_PUBLIC_METHODS,
        ALL_DECLARED_CLASSES,
        ALL_PUBLIC_CLASSES;
    }
}
`

**Comment by lbernardomaia** [comment]
Agree, but that would be a breaking change for the previous version. Is that a problem?

Removing those booleans, the annotation would be.

```
@retention(RetentionPolicy.RUNTIME)
@target(ElementType.TYPE)
public @interface RegisterForReflection {

   Attributes[] enableAttributes() default {Attributes.ALL};

   enum Attributes {
       ALL,
       ALL_DECLARED_CONSTRUCTOR,
       ALL_PUBLIC_CONSTRUCTORS,
       ALL_DECLARED_METHODS,
       ALL_PUBLIC_METHODS,
       ALL_DECLARED_CLASSES,
       ALL_PUBLIC_CLASSES;
   }
}
```

**Comment by gsmet** [comment]
I think I would keep the methods for now and mark them as deprecated.

**Comment by alexey-anufriev** [comment]
Is there any progress here? I would also like to switch to configuration-in-code.
I can also help with a contribution.

**Comment by juanmbellini** [comment]
Hey guys! I would like to add that the `fields` method could receive a list of annotation like the following:

```java
public @interface FieldData {
    String name();
    boolean allowWrite default false;
    // ...
}
```

So the `RegisterForReflection` definition could have the following method:

```java
@retention(RetentionPolicy.RUNTIME)
@target(ElementType.TYPE)
public @interface RegisterForReflection {

   FieldData[] fields;
}
```

You would use it like this:

```java
@RegisterForReflection(fields = {@FieldData(name = "id", allowWrite = true), @FieldData(name = "name", allowWrite = true), }
public class MyModel {

    private final UUID id;
    private final String name;
}
```

**Comment by gastaldi** [comment]
@juanmbellini in this case wouldn't it be better to have an annotation placed in the field itself?

**Comment by juanmbellini** [comment]
We could have both. The class level annotation can be used for registering fields of a super class that cannot be modified (for example, if it is imported from an external library)

**Comment by gastaldi** [comment]
You can specify external classes in `@RegisterForReflection` so I am not sure how would that work in this case

**Comment by juanmbellini** [comment]
@gastaldi But how would you place the field level annotation on those?

**Comment by gastaldi** [comment]
Exactly, I don't think that's going to work either in this case :)

**Comment by juanmbellini** [comment]
@gastaldi thats why I proposed using a "fields" method in the register for reflection annotation, but also a field level annotation, in order to get this feature in any of those flavors.

**Comment by juanmbellini** [comment]
I see that `RegisterForReflection` is processed `RegisterForReflectionBuildStep` and that `RegisterForReflectionBuildStep` uses `ReflectiveClassBuildItem`. 

`ReflectiveClassBuildItem` has a `Builder` which allows setting the `finalFieldsWritable` property. Couldn't we add the `finalFieldsWritable` to the `RegisterForReflection` annotation? It's not that flexible as we talked before, but at least we could start with this feature.

**Comment by yesunch** [comment]
Hello guys,I'd like to help to contribute ^^. I just saw `finalFieldsWritable` has been marked as `Deprecated`. 

So I suggest maybe we can:
-  mark the booleans `RegisterForReflection.methods`, `RegisterForReflection.fields` as `Deprecated` but keep them as @gsmet suggested
- add `Attributes[] enableAttributes()` to `RegisterForReflection` and `ReflectiveClassBuildItem`, and handle it properly in `NativeImageReflectConfigStep.generateReflectConfig`

So the `RegisterForReflection` would be:

```
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface RegisterForReflection {

   @Deprecated
    boolean methods() default true;

   @Deprecated
    boolean fields() default true;
...

   Attributes[] enabledAttributes() default {};

    enum Attributes {
        ALL_DECLARED_METHODS,
        ALL_PUBLIC_METHODS,
        ALL_DECLARED_CLASSES,
        ALL_PUBLIC_CLASSES;
    }
}
```

The `ReflectiveClassBuildItem` would be:
```
public final class ReflectiveClassBuildItem extends MultiBuildItem {

    private final List<String> className;
    private final boolean methods;
    private final boolean fields;
    private final boolean constructors;
   private final Attributes[] enabledAttributes;
   ...
}
```

What do you think about it? Any comment is appreciated!

**Comment by zakkak** [comment]
@yesunch's suggestion makes sense.

My main concern with the so far suggestions is that they seem to be tightly coupled with what's supported by GraalVM and this is a moving target (see https://github.com/oracle/graal/issues/9679).
